### PR TITLE
feat: support configurability

### PR DESCRIPTION
**Pull request by bzp2010**
Supports configuration file and environment variable configuration.

### CODE REVIEW

**Comment by bzp2010** [comment]
### Update

Configuration files and corresponding CLI flag support have been added. It uses a configuration file in the following format, one part for configuring the HTTP/grpc server (not supported by TLS yet) and another part for the DSN.

```yaml
server:
  host: 127.0.0.1
  port: 12379
# tls:
#   cert: /path/to/file.pem
#   key: /path/to/file.pem

dsn: mysql://root@tcp(127.0.0.1:3306)/apisix
```

DSN currently supports `mysql`, if not for `mysql` will default back to `btree`.

**Comment by tokers** [comment]
> ### Update
> Configuration files and corresponding CLI flag support have been added. It uses a configuration file in the following format, one part for configuring the HTTP/grpc server (not supported by TLS yet) and another part for the DSN. DSN currently supports `mysql`, if not for `mysql` will default back to `btree`.
> 
> ```yaml
> server:
>   host: 127.0.0.1
>   port: 12379
> # tls:
> #   cert: /path/to/file.pem
> #   key: /path/to/file.pem
> 
> dsn: mysql://root@tcp(127.0.0.1:3306)/apisix
> ```
> 
> This PR is currently based on the CLI branch, so it incorporates some of the original changes, and after the previous PR is merged I will do a rebase to reduce the chon

An explicit `type` field will be better than relying on the `dsn`, for btree, there is no concept of `dsn`, this is not a generic one.

File: internal/utils/log.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package utils
+
+import (
+	"go.uber.org/zap"
+)
+
+var log *zap.Logger
```
Comment (tokers) [inline internal/utils/log.go:7]: See this package: https://github.com/api7/gopkg.

**Comment by bzp2010** [comment]
### Update

These changes have been made since the last review.

- rebase to main branch
- refactor log module to reuse `api7/gopkg`
- refactor configuration

This format of the configuration file is now. The configuration format of the data source is completely modified.

```yaml
datasource:
  type: mysql
  mysql:
    host: 127.0.0.1
    port: 3306
    username: root
    password:
    database: apisix
```

File: internal/utils/log.go (hunk 1 above)
Comment (bzp2010) [reply internal/utils/log.go]: already refactor

File: internal/config/config.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	"github.com/spf13/viper"
+	"go.uber.org/zap"
+
+	"github.com/api7/etcd-adapter/internal/utils"
+)
+
+var (
+	// the DSN for database
+	DSN string
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply internal/config/config.go]: changed

**Comment by bzp2010** [comment]
ping @tokers

File: cmd/root.go [hunk 3]
```diff
@@ -58,18 +85,27 @@ var rootCmd = &cobra.Command{
...
+			}
+			logger.Info("See you next time!")
 		}
 	},
 }
 
 // Execute bootstrap root command.
 func Execute() {
+	// declare flags
+	rootCmd.PersistentFlags().StringVarP(&configFile, "config", "c", "", "config file")
```
Comment (tokers) [inline cmd/root.go:106]: We can also add the logging options.

File: internal/config/struct.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package config
+
+type server struct {
+	Host string
```
Comment (tokers) [inline internal/config/struct.go:4]: For avoiding the ambiguity from the naming style of config fields, I suggest always adding the json/yaml tag for all fields, like:

```go
Host string `json:"host" yaml:"host"`
```

Comment (bzp2010) [reply internal/config/struct.go:4]: Ok, get it. BTW, in viper it uses `mapstructure`.

Comment (bzp2010) [reply internal/config/struct.go]: fixed

File: cmd/root.go (hunk 3 above)
Comment (bzp2010) [reply cmd/root.go:106]: @tokers I found that the `api7/gopkg` log module we are using does not support setting the log level dynamically, so we cannot read the configuration after initializing the logger to set the log level. And allowing users to set log level in the CLI doesn't feel good, we even have to add parameters like log file store path. 🤔

Comment (tokers) [reply cmd/root.go:106]: A lot of programs support setting log level in the CLI, like ETCD, Ingress Controller and etc.

And the logger definitely supports setting the level dynamically. https://pkg.go.dev/github.com/api7/gopkg@v0.1.2/pkg/log#WithLogLevel

File: internal/config/struct.go (hunk 4 above)
Comment (tokers) [reply internal/config/struct.go]: Got it.

**Comment by tokers** [comment]
CI was failed @bzp2010 .

**Comment by bzp2010** [comment]
> CI was failed @bzp2010 .

Need to merge the PR for fixing CI first, I will merge it to this one branch.

**Comment by bzp2010** [comment]
## Update

After merging CLI test fix patches, the CI issue has been resolved.

ping @tokers @starsz

File: cmd/root.go (hunk 3 above)
Comment (bzp2010) [reply cmd/root.go:106]: > And the logger definitely supports setting the level dynamically. 

Yes, I noticed that, but it looks like this won't work after the logger has been created. 

The current boot process is as follows.

```text
 init logger (setup an default level logger) => init config (re-config log level and more, like log file path, but not supported; there are some log output in this stage) => start etcd-adapter
```

Currently we can define this level parameter via environment variables, but that needs to be done after parsing the configuration file normally (we can't reset the level at that time). If you think you do need to support configuring this property via the CLI, I will try to add a flag to config binding to make that.

**Review by tokers — APPROVED** [review]

**Pull request merged**